### PR TITLE
Second round of makefile/build.d consolidation

### PR DESCRIPTION
**Pull request by marler8997**
This round I'm moving the generation of the backend library to `build.d`.

### CODE REVIEW

File: src/dmd/backend/cgobj.d [hunk 1]
```diff
@@ -469,14 +469,14 @@ int insidx(char *p,uint index)
             mov     ECX,p - [ESP]           ;
         }
     else
-        asm
+        asm nothrow
```
Comment (thewilsonator) [inline src/dmd/backend/cgobj.d]: Why are these required for this PR? It feels like they should logically go in a separate one.

**Review by thewilsonator — APPROVED** [review]
Otherwise looks great!

File: src/dmd/backend/cgobj.d (hunk 1 above)
Comment (marler8997) [reply src/dmd/backend/cgobj.d]: Yeah they probably should.  This gets rid of some warnings, looks like the windows makefile ignored the warnings but the build.d isn't.  Will probably move these into another PR before this PR is ready.

**Comment by ZombineDev** [comment]
GDC9 on SemaphoreCI fails with:
```
01:17
AUTO SCROLL ONLINE WRAP ON
+ export MODEL=64
+ MODEL=64
+ export DMD=gdc
+ DMD=gdc
+ export N=4
+ N=4
+ export OS_NAME=linux
+ OS_NAME=linux
+ export FULL_BUILD=false
+ FULL_BUILD=false
+ '[' -n 10212 ']'
++ curl -fsSL https://api.github.com/repos/dlang/dmd/pulls/10212
++ jq -r .base.ref
+ BRANCH=master
+ git ls-remote --exit-code --heads https://github.com/dlang/dmd.git master
+ export BRANCH
+ source ci.sh
++ set -uexo pipefail
++ '[' -z x ']'
++ '[' -z x ']'
++ '[' -z x ']'
++ '[' -z x ']'
++ '[' -z x ']'
++ '[' -z x ']'
+++ curl --version
+++ head -n 1
++ CURL_USER_AGENT='DMD-CI curl 7.35.0 (x86_64-pc-linux-gnu) libcurl/7.35.0 OpenSSL/1.0.1f zlib/1.2.8 libidn/1.28 librtmp/2.3'
++ build_path=generated/linux/release/64
++ build_path=generated/linux/release/64
++ '[' linux == linux ']'
++ mkdir -p linker
++ rm -f linker/ld
++ ln -s /usr/bin/ld.gold linker/ld
++ NM='nm --print-size'
++ export PATH=/home/runner/dmd/linker:/home/runner/.nvm/versions/node/v8.12.0/bin:/home/runner/.rbenv/shims:/home/runner/.rbenv/bin:/home/runner/.kerl/installs/20.3/bin:/home/runner/.kiex/elixirs/elixir-1.5.3/bin:/home/runner/.kiex/bin:/home/runner/.phpbrew/php/php-5.6.37/bin:/home/runner/.phpbrew/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/jvm/java-8-oracle/bin:/usr/lib/jvm/java-8-oracle/db/bin:/usr/lib/jvm/java-8-oracle/jre/bin:/usr/lib/dart/bin/:/home/runner/.yarn/bin
++ PATH=/home/runner/dmd/linker:/home/runner/.nvm/versions/node/v8.12.0/bin:/home/runner/.rbenv/shims:/home/runner/.rbenv/bin:/home/runner/.kerl/installs/20.3/bin:/home/runner/.kiex/elixirs/elixir-1.5.3/bin:/home/runner/.kiex/bin:/home/runner/.phpbrew/php/php-5.6.37/bin:/home/runner/.phpbrew/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/jvm/java-8-oracle/bin:/usr/lib/jvm/java-8-oracle/db/bin:/usr/lib/jvm/java-8-oracle/jre/bin:/usr/lib/dart/bin/:/home/runner/.yarn/bin
+ install_d gdc
+ '[' gdc == gdc ']'
+ export DMD=gdmd-9
+ DMD=gdmd-9
+ '[' '!' -e /home/runner/dlang/gdc-9/activate ']'
+ case $1 in
+ testsuite
+ date
Wed Jul 24 05:24:39 UTC 2019
+ for step in build test rebuild '"rebuild 1"' test_dmd
+ build
+ source /home/runner/dlang/gdc-9/activate
+ make -j4 -C src -f posix.mak MODEL=64 HOST_DMD=gdmd-9 ENABLE_RELEASE=1 ENABLE_WARNINGS=1 all
no cpu specified, assuming X86
make: Entering directory `/home/runner/dmd/src'
gdmd-9 -of../generated/build -debug build.d
  (HOST_DMD_RUN)  ../config.d  ../config.d
gdmd-9 ../config.d -of../generated/linux/release/64/version_check
echo "$DEFAULT_DMD_CONF" > ../generated/linux/release/64/dmd.conf
../generated/build HOST_DMD="gdmd-9" OS=linux BUILD=release MODEL=64 AUTO_BOOTSTRAP="" --called-from-make ../generated/linux/release/64/lexer.a
../generated/build HOST_DMD="gdmd-9" OS=linux BUILD=release MODEL=64 AUTO_BOOTSTRAP="" --called-from-make ../generated/linux/release/64/backend.o
../generated/build HOST_DMD="gdmd-9" OS=linux BUILD=release MODEL=64 AUTO_BOOTSTRAP="" --called-from-make ../generated/linux/release/64/SYSCONFDIR.imp
Success
(DC) OPTABGEN optabgen.d
(RUN) OPTABBIN tytab.d
(DC) D_BACK_OBJS bcomplex.d, evalu8.d, divcoeff.d, dvec.d, go.d, gsroa.d, glocal.d, gdag.d, gother.d, gflow.d, out.d, gloop.d, compress.d, cgelem.d, cgcs.d, ee.d, cod4.d, cod5.d, nteh.d, blockopt.d, mem.d, cg.d, cgreg.d, dtype.d, debugprint.d, fp.d, symbol.d, elem.d, dcode.d, cgsched.d, cg87.d, cgxmm.d, cgcod.d, cod1.d, cod2.d, cod3.d, cv8.d, dcgcv.d, pdata.d, util2.d, var.d, md5.d, backconfig.d, ph2.d, drtlsym.d, dwarfeh.d, ptrntab.d, dvarstats.d, dwarfdbginf.d, cgen.d, os.d, goh.d, barray.d, cgcse.d, elpicpie.d, elfobj.d, aarray.d
Success
(DC) D_LEXER_OBJ console.d, entity.d, errors.d, filecache.d, globals.d, id.d, identifier.d, lexer.d, tokens.d, utf.d, array.d, ctfloat.d, file.d, filename.d, hash.d, outbuffer.d, port.d, rmem.d, rootobject.d, stringtable.d
Success
gdmd-9 -of../generated/linux/release/64/dmd -m64 -vtls -J../generated/linux/release/64 -J../res  -version=MARS -fPIC -J../generated/linux/release/64 -w -de -O -release -inline dmd/access.d dmd/aggregate.d dmd/aliasthis.d dmd/apply.d dmd/argtypes.d dmd/argtypes_sysv_x64.d dmd/arrayop.d dmd/arraytypes.d dmd/astcodegen.d dmd/ast_node.d dmd/attrib.d dmd/builtin.d dmd/canthrow.d dmd/cli.d dmd/clone.d dmd/compiler.d dmd/complex.d dmd/cond.d dmd/constfold.d dmd/cppmangle.d dmd/cppmanglewin.d dmd/ctfeexpr.d dmd/ctorflow.d dmd/dcast.d dmd/dclass.d dmd/declaration.d dmd/delegatize.d dmd/denum.d dmd/dimport.d dmd/dinifile.d dmd/dinterpret.d dmd/dmacro.d dmd/dmangle.d dmd/dmodule.d dmd/doc.d dmd/dscope.d dmd/dstruct.d dmd/dsymbol.d dmd/dsymbolsem.d dmd/dtemplate.d dmd/dversion.d dmd/escape.d dmd/expression.d dmd/expressionsem.d dmd/func.d dmd/hdrgen.d dmd/id.d dmd/impcnvtab.d dmd/imphint.d dmd/init.d dmd/initsem.d dmd/inline.d dmd/inlinecost.d dmd/intrange.d dmd/json.d dmd/lambdacomp.d dmd/lib.d dmd/libelf.d dmd/libmach.d dmd/link.d dmd/mars.d dmd/mtype.d dmd/nogc.d dmd/nspace.d dmd/objc.d dmd/opover.d dmd/optimize.d dmd/parse.d dmd/permissivevisitor.d dmd/sapply.d dmd/templateparamsem.d dmd/sideeffect.d dmd/statement.d dmd/staticassert.d dmd/target.d dmd/typesem.d dmd/traits.d dmd/transitivevisitor.d dmd/parsetimevisitor.d dmd/visitor.d dmd/typinf.d dmd/utils.d dmd/scanelf.d dmd/scanmach.d dmd/statement_rewrite_walker.d dmd/statementsem.d dmd/staticcond.d dmd/safe.d dmd/blockexit.d dmd/printast.d dmd/semantic2.d dmd/semantic3.d dmd/irstate.d dmd/toctype.d dmd/glue.d dmd/gluelayer.d dmd/todt.d dmd/tocsym.d dmd/toir.d dmd/dmsc.d dmd/tocvdebug.d dmd/s2ir.d dmd/toobj.d dmd/e2ir.d dmd/eh.d dmd/iasm.d dmd/iasmdmd.d dmd/iasmgcc.d dmd/objc_glue.d dmd/backend/cc.d dmd/backend/cdef.d dmd/backend/cgcv.d dmd/backend/code.d dmd/backend/cv4.d dmd/backend/dt.d dmd/backend/el.d dmd/backend/global.d dmd/backend/obj.d dmd/backend/oper.d dmd/backend/outbuf.d dmd/backend/rtlsym.d dmd/backend/code_x86.d dmd/backend/iasm.d dmd/backend/codebuilder.d dmd/backend/ty.d dmd/backend/type.d dmd/backend/exh.d dmd/backend/mach.d dmd/backend/mscoff.d dmd/backend/dwarf.d dmd/backend/dwarf2.d dmd/backend/xmm.d dmd/backend/dlist.d dmd/backend/melf.d dmd/backend/varstats.di dmd/root/aav.d dmd/root/array.d dmd/root/ctfloat.d dmd/root/file.d dmd/root/filename.d dmd/root/man.d dmd/root/outbuffer.d dmd/root/port.d dmd/root/response.d dmd/root/rmem.d dmd/root/rootobject.d dmd/root/speller.d dmd/root/longdouble.d dmd/root/strtold.d dmd/root/stringtable.d dmd/root/hash.d dmd/root/string.d ../generated/linux/release/64/lexer.a ../generated/linux/release/64/backend.o
In function ‘path’,
    inlined from ‘ensurePathExists’ at dmd/root/filename.d:862:22:
dmd/root/filename.d:325:9: error: ‘__builtin_memcpy’ specified bound 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
  325 |         memcpy(path, str.ptr, pathlen);
      |         ^
In function ‘path’,
    inlined from ‘ensurePathToNameExists’ at dmd/utils.d:100:18:
dmd/root/filename.d:325:9: error: ‘__builtin_memcpy’ specified bound 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
  325 |         memcpy(path, str.ptr, pathlen);
      |         ^
d21: all warnings being treated as errors
make: *** [../generated/linux/release/64/dmd] Error 1
```

File: src/build.d [hunk 2]
```diff
@@ -755,22 +781,39 @@ auto sourceFiles()
...
+            ~ versionTernary!"Windows"("cgobj.d filespec.d mscoffobj.d newman.d".split, ["aarray.d"])
+        ).map!(e => env["C"].buildPath(e)).array,
+        backendHeaders: "
+            cc.d cdef.d cgcv.d code.d cv4.d dt.d el.d global.d
+            obj.d oper.d outbuf.d rtlsym.d code_x86.d iasm.d codebuilder.d
+            ty.d type.d exh.d mach.d mscoff.d dwarf.d dwarf2.d xmm.d
+            dlist.d melf.d varstats.di
+        ".split.map!(e => env["C"].buildPath(e)).array,
     };
+    sources.backend ~=  ((env["OS"] == "windows" && env["MODEL"] == "64") ? ["strtold.d", "longdouble.d"] : [])
```
Comment (wilzbach) [inline src/build.d]: Why not use a good old if?

File: src/build.d [hunk 3]
```diff
@@ -755,22 +781,39 @@ auto sourceFiles()
...
+
     sources.dmd = sources.frontend ~ sources.backendHeaders;
 
     return sources;
 }
 
+/// Returns the first argument if the given version is true, otherwise, returns the second argument.
+auto versionTernary(string version_, T)(T trueCase, T falseCase)
+{
+    mixin("version(" ~ version_ ~ ") { return trueCase; }");
```
Comment (wilzbach) [inline src/build.d]: Can't we look up the OS in `env`?

File: src/build.d [hunk 4]
```diff
@@ -731,16 +758,23 @@ auto sourceFiles()
...
-        ].map!(e => env["C"].buildPath(e ~ ".d")).array,
+        backend: ("
+            bcomplex.d evalu8.d divcoeff.d dvec.d go.d gsroa.d glocal.d gdag.d gother.d gflow.d
+            out.d
+            gloop.d compress.d cgelem.d cgcs.d ee.d cod4.d cod5.d nteh.d blockopt.d mem.d cg.d cgreg.d
+            dtype.d debugprint.d fp.d symbol.d elem.d dcode.d cgsched.d cg87.d cgxmm.d cgcod.d cod1.d cod2.d
+            cod3.d cv8.d dcgcv.d pdata.d util2.d var.d md5.d backconfig.d ph2.d drtlsym.d dwarfeh.d ptrntab.d
+            dvarstats.d dwarfdbginf.d cgen.d os.d goh.d barray.d cgcse.d elpicpie.d
+            ".split
+            ~ ( (env["OS"] == "osx") ? ["machobj.d"] : ["elfobj.d"] )
```
Comment (wilzbach) [inline src/build.d]: I presume you copied this from the posix makefil, but as there's windows too shouldn't it be sth. like:

```
( (env["OS"] == "osx") ? ["machobj.d"] : [] )
( (env["OS"] == "linux" || env["OS"] == "freebsd") ? ["elfobj.d"] : [] )
```

Though you can skip all of this logic, because these files have `version (OS)` headers, so there's no problem in keeping things simple and including all.

**Comment by wilzbach** [comment]
Windows failure on Azure is not very insightful:

```
Test runnable\test18772.d failed.  The logged output:
..\generated\windows\release\64\dmd.exe -conf= -m64 -Irunnable   -L/OPT:NOICF  -odtest_results\runnable -oftest_results\runnable\test18772_0.exe  runnable\test18772.d  -map nul.map


Assertion failed: stackused == 0, file D:\a\1\s\src\dmd\backend\cgcod.d, line 697


==============================

Test runnable\test18772.d failed: expected rc == 0, exited with rc == 9
```

CC @rainers

**Review by wilzbach — COMMENTED** [review]
.

Edit: GH mobile false review.

**Comment by ZombineDev** [comment]
This PR broke CircleCI for phobos. The difference between the dmd CircleCI build and the phobos CircleCI build comes down to:
* `make -f posix.mak HOST_DMD=dmd` (dmd)
* `make -f posix.mak HOST_DMD=dmd BUILD=debug all` (phobos)

I.e. the `BUILD=debug` build type is broken. Compare:
```sh
git checkout 1309c79ec # The merge commit of this PR
make -f posix.mak HOST_DMD=dmd clean
make -f posix.mak HOST_DMD=dmd BUILD=debug all # fails

git checkout 1309c79ec^ # Checkout dmd before this PR was merged
make -f posix.mak HOST_DMD=dmd clean
make -f posix.mak HOST_DMD=dmd BUILD=debug all # succeeds
```
